### PR TITLE
deprecated Twig_ExtensionInterface::initRuntime()

### PR DESCRIPTION
**Pull request by fabpot**
I'm working on splitting Twig extensions into 2 different phases: compilation and runtime. The goal is to avoid having to load the runtime environment of an extension when compiling templates. That also opens the way to be able to lazy-load Twig extensions.

While working on the split, I realized that the `initRuntime()` method on `Twig_ExtensionInterface` is not needed anymore. It was added at a time `needs_environment` did not exist and was a way to keep the environment around for custom filters/tests/functions. But nowadays, that's not needed anymore. I did a quick search on Github, and most of the implementation I found just store the environment in a local property, which is not needed anymore. So, I propose to deprecate it in 1.x and remove it in 2.0.

### CODE REVIEW

**Comment by fabpot** [comment]
And of course, I've just had a look at Symfony, and we are using this method in the Twig form extension. I need to see why and how to fix that.

**Comment by fabpot** [comment]
Ok, it's probably to avoid a circular reference, we can probably find another way.

File: test/Twig/Tests/EnvironmentTest.php [hunk 1]
```diff
@@ -54,11 +54,11 @@ public function testGlobals()
         $globals = $twig->getGlobals();
         $this->assertEquals('bar', $globals['foo']);
 
-        // globals can be modified after runtime init
+        // globals can be modified after a template has been loaded
         $twig = new Twig_Environment($this->getMock('Twig_LoaderInterface'));
         $twig->addGlobal('foo', 'foo');
         $twig->getGlobals();
-        $twig->initRuntime();
+        $twig->loadTemplate('index');
```
Comment (fabpot) [inline test/Twig/Tests/EnvironmentTest.php]: This is more semantically correct. Calling `initRuntime` is just a side-effect of loading a template.

**Comment by stof** [comment]
@fabpot I suggest finding the solution in Symfony before merging this

**Comment by fabpot** [comment]
To keep BC, I've added a new interface: `Twig_Extension_InitRuntimeInterface` which can be implemented if really needed (see changes for more information).

File: doc/deprecated.rst [hunk 2]
```diff
@@ -26,6 +26,13 @@ Extensions
 * As of Twig 1.x, the ability to remove an extension is deprecated and the
   ``Twig_Environment::removeExtension()`` method will be removed in 2.0.
 
+* As of Twig 1.x, the ``Twig_ExtensionInterface::initRuntime()`` is deprecated.
```
Comment (stof) [inline doc/deprecated.rst]: Please say ``As of Twig 1.23``. ``As of 1.x`` does not make sense as ``1.x`` is not a version

File: lib/Twig/Extension/InitRuntimeInterface.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Declares the deprecated Twig_Extension::initRuntime() method.
+ *
+ * Implement explicitely this interface if you really need to implement the
```
Comment (xabbuh) [inline lib/Twig/Extension/InitRuntimeInterface.php]: Explicitly implement this [...]

File: doc/deprecated.rst (hunk 2 above)
Comment (xabbuh) [inline doc/deprecated.rst]: [...] the `Twig_ExtensionInterface::initRuntime()` method is [...]

Comment (fabpot) [inline doc/deprecated.rst]: fixed typo
@stof we are not using that anywhere, that should be done in a separate PR.

File: lib/Twig/Extension/InitRuntimeInterface.php (hunk 3 above)
Comment (fabpot) [inline lib/Twig/Extension/InitRuntimeInterface.php]: fixed

File: lib/Twig/Extension/InitRuntimeInterface.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+interface Twig_Extension_InitRuntimeInterface
+{
+    /**
+     * Initializes the runtime environment.
+     *
+     * This is where you can load some file that contains filter functions for instance.
+     *
+     * @param Twig_Environment $environment The current Twig_Environment instance
+     */
+    public function initRuntime(Twig_Environment $environment);
```
Comment (stof) [inline lib/Twig/Extension/InitRuntimeInterface.php]: This interface can be implemented at the same time than Twig_ExtensionInterface only in PHP 5.3.9+, as it defines the same method (remember our issues with the Validator component in Symfony 2.5).
So this makes it unsuited for the Twig 1.x lowest bound (and for usage in Symfony 2.3): https://3v4l.org/PQSTH

Comment (fabpot) [inline lib/Twig/Extension/InitRuntimeInterface.php]: So, what about just making this interface a simple marker?

File: doc/deprecated.rst (hunk 2 above)
Comment (stof) [inline doc/deprecated.rst]: Well, using the exact version for the new one already makes it simpler as we know which version it is.

File: lib/Twig/Extension/InitRuntimeInterface.php (hunk 4 above)
Comment (fabpot) [inline lib/Twig/Extension/InitRuntimeInterface.php]: done

Comment (stof) [inline lib/Twig/Extension/InitRuntimeInterface.php]: I suggest adding a test covering the usage of this interface in the Twig testsuite

Comment (fabpot) [inline lib/Twig/Extension/InitRuntimeInterface.php]: done

File: test/Twig/Tests/EnvironmentTest.php [hunk 5]
```diff
@@ -308,6 +308,33 @@ public function testAddMockExtension()
...
+    public function testInitRuntimeWithAnExtensionUsingInitRuntimeDeprecation()
+    {
+        $twig = new Twig_Environment($this->getMock('Twig_LoaderInterface'));
+        $twig->addExtension(new Twig_Tests_EnvironmentTest_ExtensionWithDeprecationInitRuntime());
+
+        $deprecations = array();
+        set_error_handler(function ($type, $msg) use (&$deprecations) {
+            if (E_USER_DEPRECATED === $type) {
+                $deprecations[] = $msg;
+            }
```
Comment (stof) [inline test/Twig/Tests/EnvironmentTest.php]: you should call the PHPUnit error handler for any other error, to avoid loosing them

Comment (fabpot) [inline test/Twig/Tests/EnvironmentTest.php]: Instead, I've added an assertion on the deprecation message.

**Comment by fabpot** [comment]
Not sure why it does not work on 5.2.
